src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -3582,15 +3582,18 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
 
         case _ if currentRun.runDefinitions.isPolymorphicSignature(fun.symbol) =>
           // Mimic's Java's treatment of polymorphic signatures as described in
-          // https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.12.3
+          // https://docs.oracle.com/javase/specs/jls/se11/html/jls-15.html#jls-15.12.3
           //
           // One can think of these methods as being infinitely overloaded. We create
           // a fictitious new cloned method symbol for each call site that takes on a signature
           // governed by a) the argument types and b) the expected type
           val args1 = typedArgs(args, forArgMode(fun, mode))
           val clone = fun.symbol.cloneSymbol.withoutAnnotations
           val cloneParams = args1.map(arg => clone.newValueParameter(freshTermName()).setInfo(arg.tpe.deconst))
-          val resultType = if (isFullyDefined(pt)) pt else ObjectTpe
+          val resultType =
+            if (fun.symbol.tpe.resultType.typeSymbol != ObjectClass) fun.symbol.tpe.resultType
+            else if (isFullyDefined(pt)) pt
+            else ObjectTpe
```

Comment:
Thanks for this, I'd just noticed a problem here yesterday!

Before:
```
scala> import java.lang.invoke._
import java.lang.invoke._

scala> class Data { var a: Int = 0; def aHandle = MethodHandles.lookup().findVarHandle(getClass, "a", classOf[Int])}
defined class Data

scala>

scala> val d = new Data
d: Data = Data@15c4b1a4

scala>  d.aHandle.set(d, 2)
java.lang.NoSuchMethodError: VarHandle.set(Data,int)Object
  at java.base/java.lang.invoke.MethodHandleNatives.newNoSuchMethodErrorOnVarHandle(MethodHandleNatives.java:589)
  at java.base/java.lang.invoke.MethodHandleNatives.varHandleOperationLinkerMethod(MethodHandleNatives.java:542)
  at java.base/java.lang.invoke.MethodHandleNatives.linkMethodImpl(MethodHandleNatives.java:475)
  at java.base/java.lang.invoke.MethodHandleNatives.linkMethod(MethodHandleNatives.java:463)
  ... 28 elided

scala>  d.aHandle.set(d, 2): Unit // workaround, fully define expected type as Unit

scala> d.a
res2: Int = 2
```

With your PR:
```

```
scala> d.aHandle.set(d, 2)

scala> d.a
val res9: Int = 2
```